sycl/doc/extensions/ExplicitSIMD/dpcpp-explicit-simd.md

```diff
@@ -0,0 +1,776 @@
+# Explicit SIMD Programming Extension for DPC++
+
+## Introduction
+
+The main motivation for introducing the Explicit SIMD Programming (ESP) DPC++
+extension is enabling low-level efficient programming for Intel Gen
+architectures. More specifically, Explicit SIMD provides the following
+additional features:
+- Manual vectorization of device code using the `simd` class mapped to Gen's
+  general register file. This allows to write efficient code not relying on
+  further widening by the compiler, as with traditional SPMD programming.
+- Low-level APIs efficiently mapped to Gen architecture, such as block reads.
+
+Here are some future directions in which this API is intended to evolve:
+- enabling this extension for other architectures, such as x86, with extracting
+  and clearly marking generic and target-dependent API portions
+- aligning with `std::simd` and maybe providing `std::simd` implementation atop
```

Comment:
to avoid re-implementing lots of stuff and take advantage of sycl::intel::gpu::simd efficiency

Reply:
Implementing it on top of `gpu::simd` is the strategy which requires a reimplementation of `std::simd` features missing from `gpu::simd`. On the other hand there is an existing implementation directly on top of gcc/clang vector: https://github.com/VcDevel/std-simd . We should explore how we can reuse most of that implementation also for the GPU.

Reply:
This is a pretty fundamental design question, I'm sorry if I am missing context, but I just found out your work today. First of all, :+1:  thanks for working on a SIMD type for GPU. I wanted to do this since years but didn't have the manpower/time.

What I'd like to know is why you can't (don't want to) take the `std::experimental::simd` specification as is. I'd be very happy to help. Especially with folding back improvements and extensions into C++ committee papers targeting C++23. If the `simd<T, ?>` type is 1:1 portable between CPU and GPU that would be a huge win in terms of software maintenance, and freedom to move to newer platforms.

Reply:
The short answer is that the primary target for this specification is Intel GPU, which is reflected in the namespace.

But generalizing the spec (and implementation) to fit more devices and be more compatible
with std C++ is definitely welcome, as long as it does not hurt GPU performance. simd<T, ?>
type is portable to CPU, but performance of various associated APIs is another question. Our
plan is to upstream the existing initial implementation, then iteratively improve it along with
the specification, making sure it remains functional and performant on Intel GPU.

Reply:
> The short answer is that the primary target for this specification is Intel GPU, which is reflected in the namespace.

Right. I think it totally makes sense to explore what would be the perfect API for programming a specific target. It makes less sense to write applications with it. Because portability and being able to move to newer/other hardware is an important reason for using SIMD types in the first place.
Also, "Intel GPU" is a narrow target right now. But do you know whether it will stay that way?

> But generalizing the spec (and implementation) to fit more devices and be more compatible
> with std C++ is definitely welcome, as long as it does not hurt GPU performance.

:+1: A SIMD type that doesn't allow you to reach full performance is a failure (or hopefully just buggy and needs to be fixed). Performance is the reason why we use it.

I guess what I'd like to discuss is: *If you deviate from the Parallelism TS 2, why? Is it to simplify your implementation, is it because of missing functionality, is it because of performance, or?* This is important feedback to the C++ committee when merging the TS into the IS is discussed.

Reply:
To help you understand our points of view, I recommend you take a look the documents and examples at https://github.com/intel/cm-compiler

Reply:
Be sure that all the interesting links which were mentioned in the discussions end up in the extension description itself, so it is not lost for the future.

Reply:
Added these comments to the bottom of the doc as TODO